iroh/src/magicsock.rs

```diff
@@ -692,77 +753,83 @@ impl MagicSock {
+                if datagram.len() < meta.stride {
                     trace!(
-                        len = %packet.len(),
+                        len = %datagram.len(),
                         %meta.stride,
                         "Last GRO datagram smaller than stride",
                     );
                 }
 
-                let packet_is_quic = if stun::is(packet) {
+                // Detect DISCO and STUN datagrams and process them.  Overwrite the first
+                // byte of those packets with zero to make Quinn ignore the packet.  This
+                // relies on quinn::EndpointConfig::grease_quic_bit being set to `false`,
+                // which we do in Endpoint::bind.
+                if stun::is(datagram) {
                     trace!(src = %meta.addr, len = %meta.stride, "UDP recv: stun packet");
-                    let packet2 = Bytes::copy_from_slice(packet);
+                    let packet2 = Bytes::copy_from_slice(datagram);
                     self.net_reporter.receive_stun_packet(packet2, meta.addr);
-                    false
-                } else if let Some((sender, sealed_box)) = disco::source_and_box(packet) {
-                    // Disco?
+                    datagram[0] = 0u8;
+                } else if let Some((sender, sealed_box)) = disco::source_and_box(datagram) {
                     trace!(src = %meta.addr, len = %meta.stride, "UDP recv: disco packet");
                     self.handle_disco_message(
                         sender,
                         sealed_box,
                         DiscoMessageSource::Udp(meta.addr),
                     );
-                    false
+                    datagram[0] = 0u8;
                 } else {
                     trace!(src = %meta.addr, len = %meta.stride, "UDP recv: quic packet");
                     if from_ipv4 {
-                        inc_by!(MagicsockMetrics, recv_data_ipv4, packet.len() as _);
+                        inc_by!(MagicsockMetrics, recv_data_ipv4, datagram.len() as _);
                     } else {
-                        inc_by!(MagicsockMetrics, recv_data_ipv6, packet.len() as _);
+                        inc_by!(MagicsockMetrics, recv_data_ipv6, datagram.len() as _);
                     }
-                    true
+                    quic_datagram_count += 1;
+                    buf_contains_quic_datagrams = true;
                 };
-
-                if packet_is_quic {
-                    quic_packets_count += 1;
-                    is_quic = true;
-                } else {
-                    // overwrite the first byte of the packets with zero.
-                    // this makes quinn reliably and quickly ignore the packet as long as
-                    // [`quinn::EndpointConfig::grease_quic_bit`] is set to `false`
-                    // (which we always do in Endpoint::bind).
-                    packet[0] = 0u8;
-                }
```

Comment:
I guess this is subjective.  I think the way I wrote it has less cognitive overhead since you save a variable which you have to store on your mental stack of things to keep track off and link back to the source.  There is also something about assignment to a variable from 3 different places with like a dozen lines of code between them that I find hard to visually keep track off to map back to the variable.

And since the thing you're doing for each branch is only a single line I prefer this version and don't find the extra cognitive load of a variable worthwhile  :shrug: